test/jdk/java/awt/ScrollPane/ScrollPaneRemoveAdd.java

```diff
@@ -0,0 +1,168 @@
+
+public class ScrollPaneRemoveAdd {
+    Button button;
+    ScrollPane pane;
+    Semaphore actionSema;
+    Frame frame;
+    Robot robot;
+    volatile Point buttonLoc;
+    volatile Dimension buttonSize;
+
+    public static void main(String[] args) throws Exception {
+        ScrollPaneRemoveAdd scrollTest = new ScrollPaneRemoveAdd();
+        scrollTest.init();
+        scrollTest.start();
+    }
+
+    public void init() throws Exception {
+        EventQueue.invokeAndWait(() -> {
+            frame = new Frame("Scroll pane Add/Remove");
+
+            pane = new ScrollPane(ScrollPane.SCROLLBARS_ALWAYS);
+            button = new Button("press");
+            pane.add(button);
+            frame.add(pane);
+            actionSema = new Semaphore();
+            button.addActionListener(new ActionListener() {
+                public void actionPerformed(ActionEvent e) {
+                    actionSema.raise();
+                }
+            });
+        });
+        robot = new Robot();
+    }
+
+    public void start() throws Exception {
+        try {
+            EventQueue.invokeAndWait(() -> {
+                frame.pack();
+                frame.setLocationRelativeTo(null);
+                frame.setVisible(true);
+                pane.remove(0);
+                pane.add(button);
+                buttonLoc = button.getLocationOnScreen();
+                buttonSize = button.getSize();
+            });
+
+            robot.mouseMove(buttonLoc.x + buttonSize.width / 2,
+                    buttonLoc.y + buttonSize.height / 2);
+            robot.delay(50);
+            robot.mousePress(InputEvent.BUTTON1_DOWN_MASK);
+            robot.delay(50);
+            robot.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
+
+            robot.delay(50);
+            robot.waitForIdle();
+            try {
+                actionSema.doWait(1000);
+            } catch (Exception ie) {
+                throw new RuntimeException("My sleep was interrupted");
+            }
+            robot.delay(50);
+            robot.waitForIdle();
+            if (!actionSema.getState()) {
+                throw new RuntimeException("ScrollPane doesn't handle " +
+                        "correctly add after remove");
+            }
+        } finally {
+            EventQueue.invokeAndWait(() -> {
+                if (frame != null) {
+                    frame.dispose();
+                }
+            });
+        }
+    }
+
+    private static class Semaphore {
+        volatile boolean state = false;
+        final Object lock = new Object();
+        volatile int waiting = 0;
+
+        public Semaphore() {
+        }
+
+        public void doWait() throws InterruptedException {
+            synchronized (lock) {
+                if (state) {
+                    return;
+                }
+                waiting++;
+                synchronized (this) {
+                    wait();
+                }
+                waiting--;
+            }
+        }
```

Comment:
This method is never used, it can safely be dropped.

Reply:
Updated.